MAINT: backports for 1.15.2

tylerjereddy opened the pull request:
Merge conflicts seem to be causing many (236) test failures--I may back out PRs that had merge conflicts and that don't look critical if I don't see a straightforward path forward for them. Some cases had too much drift relative to `main` and didn't really seem like serious things to fix anyway.

Backports included so far (merge conflicts noted):

1. gh-22038
2. gh-22080
3. gh-22187
4. gh-22322
5. gh-22337
6. gh-22339
7. gh-22353
8. gh-22356
9. gh-22357
10. gh-22372
11. gh-22373
12. gh-22374
13. gh-22405 (merge conflicts)
14. gh-22418
15. gh-22423
16. gh-22446 (there may be a flaky 3.10 CI job?)
17. gh-22447 (merge conflicts)
18. gh-22472
19. gh-22482
20. gh-22402
21. gh-22408
22. gh-22345


TODO:

- [x] fix or back out patches that had merge conflicts and are causing problems
- [x] test failure `test_quadrature.py::TestCumulativeSimpson::test_cumulative_simpson_against_simpson_with_default_dx`
- [x] start restoring the CI and get it passing
- [x] pull in more desired backports (gh-22402, others?)
- [x] update the release notes and probably mention new 3.13t wheels on Linux ARM
- [x] wheel builds succeeding?
- [ ] is the CuPy team happy re: https://github.com/cupy/cupy/issues/8866 ?

andfoy committed:
BUG: immortalize uarray global strings in order to prevent negative refcounts (#22038)

#### Reference issue
Fixes https://github.com/scipy/scipy/issues/21214
Closes https://github.com/scipy/scipy/pull/21218

#### What does this implement/fix?
This PR addresses the segfaults caused by the occurrence of negative refcounts when uarray static strings were being released when the interpreter didn't exist.

---------

Co-authored-by: peterbell10 <peterbell10@live.co.uk>

mdhaber committed:
MAINT: stats.Mixture: make return type consistent

cbouss committed:
do not check dtype in test_compare_with_GCVSPL

WarrenWeckesser committed:
BUG: special: Fix unchecked memory allocations in `specfun.h` (#22080)

* WIP: Fix unchecked memory allocation in aswfa()

Summary of the changes:

* Add another SF error code, "SF_ERROR_NOMEM", that is used by ufuncs
  that require internally allocated memory.  Just like the other possible
  errors, how this error is handled can be queried and set using `geterr()`
  and `seterr()`.  Unlike the other error conditions, the default behavior
  is to raise an exception if a memory allocation fails.

* Modify the function `aswfa()` in `specfun.h` to return a status code
  that indicates when a mempory allocation failed.

* Modify the functions `prolate_aswfa_nocv()`, `oblate_aswfa_nocv()`,
  `prolate_aswfa()` and `oblate_aswfa()`  in `sphd_wave.h` to check
  the return value of `aswfa()`, and set the error state and return values
  appropriately if `aswfa()` returns `NoMemory`.

* MAINT: Rename 'nomem' -> 'memory'.

* Fix all uses of unchecked allocations in specfun.h

* Fix aswfa() in specfunc.h

* Fix refguide check.

* Add comment about the return value when nonconvergence is detected in rmn2l.

* Change (c/m)alloc calls to use 'new' wrapped in a unique_ptr.

* Fix a few more names passed to set_error()

* Set outputs to nan in mtu12 when alloc fails
Also update docstring-like comments of some C++ functions.

j-bowhay committed:
TST: interpolate: small tolerance bump to TestAAA.test_basic_functions

HaoZeke committed:
MAINT: Stop installing libhighs

Co-authored-by: drew-parsons <drew-parsons@users.noreply.github.com>

Closes #22349

dschult committed:
BUG: sparse.linalg.norm: fix return type (#22372)

cupy reports https://github.com/cupy/cupy/issues/8866 that
sparse.linalg.norm returns a python number instead of a 0-dim numpy array

j-bowhay committed:
BUG: special: Fix unchecked malloc in stirling2.h (#22339)

Use std::unique_ptr and new (std::nothrow) instead of malloc.

Closes gh-22336.

dschult committed:
DOC: sparse.linalg: add two recent functions to namespace and fix doctests (#22374)

* fix doctests and make two recent functions visible

* update refguide

[docs only]

* fix capitalization   [docs only]

Co-authored-by: Daniel Schmitz <40656107+dschmitz89@users.noreply.github.com>

---------

Co-authored-by: Daniel Schmitz <40656107+dschmitz89@users.noreply.github.com>

dschult committed:
revert NotImplemented return values in dot & others (#22373)

rgommers committed:
CI: build Linux aarch64 wheels on GitHub Actions instead of Cirrus CI

This unifies the wheels builds, which is a lot easier on maintenance.

In the process, did the following:
- update to setup-miniconda 3.1.1. for GHA Linux arm64 runner support
- add `conda-remove-defaults` to `setup-miniconda`

mdhaber committed:
MAINT: integrate.cumulative_simpson: bump test tolerance

mdhaber committed:
MAINT: stats.zmap: restore support for complex data (#22405)

* MAINT: stats.zmap: restore support for complex data

* MAINT: stats.zmap: fix masked array support

* Apply suggestions from code review

PTNobel committed:
BUG: sparse: fix selecting wrong dtype for coo coords (#22353)

* Fixes bug with selecting wrong dtype for coo coords

* add test for reshape having wrong dtype

* test that smaller dtype is maintained even if intermediate values are big

---------

Co-authored-by: Dan Schult <dschult@colgate.edu>

mdhaber committed:
MAINT: stats.Mixture: fix inverse functions when mean is undefined (#22337)

* MAINT: stats.Mixture: fix inverse when mean is undefined

* Apply suggestions from code review

* Update scipy/stats/_distribution_infrastructure.py

tylerjereddy committed:
BUG: cluster: `cophenet` intercept invalid linkage matrix count (#22187)

* BUG: catch invalid linkage count

* Fixes gh-22183.

* The input data from gh-22183 causes an out
of bounds memory access in a 1-D `memoryview`
in the Cython `cophenetic_distances` function.
Prevent this by enforcing a check for an allowable
upper bound on the cluster membership (4th column)
of the linkage matrix `Z` received by `is_valid_linkage`.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

* TST: PR 22187 revisions

* Adjust `test_gh_22183()` such that the linkage matrix is invalid
only because of an excessive observation count, rather than
also being invalid because the large observation count is
not integral.

* Adjust `test_gh_22183()` such that is properly flushes
through array API backends.

* DOC: backticks

---------

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>
Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

WarrenWeckesser committed:
BUG: special: Fix a memory leak in the AMOS function besy(). (#22423)

Use a unique_ptr to manage the heap-allocated array.

Closes gh-22314.

rgommers committed:
CI: migrate Linux aarch64 job to GitHub Actions

rgommers committed:
CI: remove Cirrus CI configuration [wheel build]

matthew-brett committed:
BUG: io.loadmat: throw error on file containing all zeros (#22447)

[ci skip]

scottshambaugh committed:
BUG: scipy.spatial: Fix inaccurate orthonormalization in `Rotation.from_matrix()` (#22418)

* BUG: fix inaccurate orthonormalization in Rotation.from_matrix()

* Code review updates

---------

Co-authored-by: Scott Shambaugh <scottshambaugh@users.noreply.github.com>

mdhaber commented:
LMK if you want help with gh-22405.

tylerjereddy committed:
BUG: PR 22471 revisions

* Fixes issues related to backport manual merge conflict resolution
during the preparation of SciPy `1.15.2`.

* Restore an accidentally-removed size-0 array check in
`_contains_nan()` function--this allows SciPy `stats` tests
to pass again on the maintenance branch.

* Remove some accidentally-added tests in `test_mio.py`

[skip circle]

tylerjereddy commented:
I identified my own errors in the manual resolution of backport merge conflicts and the full suite passes locally on ARM Mac. For now, I'll flush the regular CI but not "circle" yet since the release notes/docs are not updated yet. We'll see if the CI catches anything I don't see locally.

tylerjereddy commented:
Ok, the original 236 test failures have been fixed, but of course there is now a new one, `test_cumulative_simpson_against_simpson_with_default_dx`. For now I've added that to the TODO list above. That's brand new, I wonder if it may be showing up on `main` as well.

dschult commented:
Please consider #22472 for backporting. It fixes a bug with indexing spmatrix and using `None` in the index.
It is not merged, but hopefully that will happen before this is ready. 
Thanks!

tylerjereddy committed:
MAINT: pearsonr SIMD-related shim

* Fixes #22479.

* This small patch allows `test_stats.py::TestRegression::test_regressZEROX`
to pass on x86_64 Linux instead of failing due to an extra division by
zero warning over a fairly narrow range of NumPy versions near `1.25.2`
where SIMD implementation details appear to have been slightly
different.

[skip circle]

dschult committed:
BUG: sparse: fix spmatrix indexing with None and implicit padding to match matrix behavior (#22472)

* fix index using None when index is padded with :

* align None indexing for spmatrix with np.matrix behavior. And tests.

* rewrite spmatrix handling to ease reading

dschult commented:
Just an update that #22472 has now been merged and is ready to try backporting if/when you are able.
Thanks!

tylerjereddy committed:
DOC: PR 22471 revisions

* Update the SciPy `1.15.2` release notes following backport
activity.

tylerjereddy commented:
Ok, we're back at parity with all merged PRs with a backport label. I'm going to cross my fingers that `TestCumulativeSimpson` `hypothesis`-based test failure goes away. It wasn't showing up locally with a typical test incantation, so if it goes away in CI now I'll probably just check that item off.

I also updated the `1.15.2` release notes, including a brief mention of the new Linux ARM free-threaded `3.13` wheels.

I'll see where the CI stands after this latest push. There's still one backport not yet merged, mentioned in the TODO list above. I've been hesitant to merge that, even as release manager, since I'm the author, but it may close 4-5 issues so I may consider doing it anyway if no objections pop up soon.

tylerjereddy commented:
`test_cumulative_simpson_against_simpson_with_default_dx` is still happening; will probably need to try to reproduce locally then

tylerjereddy committed:
BUG: wrap median_filter stability (#22402)

* An uninitialized data structure in `_rank_filter()` C++ code
was causing stochastic algorithm failures on Linux but not MacOS,
so force a zero-based initialization. This allows `test_gh_22250`
to pass over thousands of `pytest-repeat` incantations.

* Add a regression test for gh-22333, which reliably passes
on this branch with `pytest-repeat` (and fails on `main` with
sufficient repeats).

tylerjereddy committed:
CI: PR 22471 revisions

* temporarily disable testing against `meson` `master` branch
in CI because of gh-22534.

tylerjereddy committed:
DOC: PR 22471 revisions

* Update the SciPy `1.15.2` release notes following
additional backport activity.

tylerjereddy commented:
Regular CI is all green now--proceeding with full wheel build matrix to assess potential flakiness of some jobs.

tylerjereddy committed:
CI: PR 22471 wheel builds [wheel build]

* This is an empty commit to test the wheel build
matrix in PR 22471.

[wheel build]

tylerjereddy commented:
There are two failing wheel builds in the first attempt:

1. Python 3.10 (musl)Linux ARM. For this one, I believe Ralf already indicated that it could be flaky, and the failure looks like something that might resolve with a restart, `Warning: error    libmamba Error when extracting package: Could not stat info/recipe/parent/patches/mingw/13/2000-enable-rust.patch`.
2. Python 3.10 MacOSX x86_64 OpenBLAS. Looks like a segfault happens in `scipy/signal/tests/test_bsplines.py` `test_sepfir2d_strided_3`. I don't see any hits in our issue or PR lists.

As a first thing to try, I'll see if restarting those two failing jobs is helpful.

ev-br commented:
> Python 3.10 MacOSX x86_64 OpenBLAS. Looks like a segfault happens in scipy/signal/tests/test_bsplines.py test_sepfir2d_strided_3. I don't see any hits in our issue or PR lists.

The test is decorated with `@pytest.mark.xfail(reason="XXX: flaky. pointers OOB on some platforms")`, 
https://github.com/scipy/scipy/blob/v1.15.1/scipy/signal/tests/test_bsplines.py#L285

It's fine to make it `@pytest.mark.skip` instead. Let me know if you want me to send a PR (to main or to the maintenance branch?)

tylerjereddy commented:
Thanks Evgeni--if it happens again here I'll just push in a commit with you as co-author (share the blame :)) to effectively suppress that one test. Perhaps keeping `xfail` but using the more aggressive `run=False` could suit to avoid execution entirely if it can crash.

tylerjereddy commented:
The CI is all green now, including the full wheel build matrix. The TODO items above are all dealt with except for the CuPy team thing, but it looks like they are happy with the latest support Lucas provided over there, so I think I can move forward with the `1.15.2` release process.
